File: javascript/dsa/algorithms/sorting/first-missing-positive.js
function firstMissingPositive(nums) {
  const n = nums.length;
  for (let i = 0; i < n; ) {
    const targetIdx = nums[i] - 1;
    if (nums[i] > 0 && nums[i] <= n && nums[targetIdx] !== nums[i]) {
      [nums[i], nums[targetIdx]] = [nums[targetIdx], nums[i]];
    } else {
      i++;
    }
  }
  for (let i = 0; i < n; i++) {
    if (nums[i] !== i + 1) {
      return i + 1;
    }
  }
  return n + 1;
}

File: cpp/dsa/algorithms/sorting/first-missing-positive.cpp
int firstMissingPositive(vector<int>& nums) {
    int n = nums.size();
    for (int i = 0; i < n; ) {
        int targetIdx = (long long)nums[i] - 1;
        if (nums[i] > 0 && nums[i] <= n && nums[targetIdx] != nums[i]) {
            int temp = nums[i];
            nums[i] = nums[targetIdx];
            nums[targetIdx] = temp;
        } else {
            i++;
        }
    }
    for (int i = 0; i < n; i++) {
        if (nums[i] != i + 1) {
            return i + 1;
        }
    }
    return n + 1;
}